src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -568,7 +568,7 @@ def expectation_from_samples(self, freq: dict, qubit_map: list = None) -> float:
 
         keys = list(freq.keys())
         counts = list(freq.values())
-        counts = self.backend.cast(counts, dtype=type(counts[0])) / sum(counts)
+        counts = self.backend.cast(counts, dtype=self.backend.np.float64) / sum(counts)
```

Comment:
```suggestion
        counts = self.backend.cast(counts, dtype=self.backend.dtype) / sum(counts)
```
This should work as well, right?

Reply:
unfortunately I think that the default dtype is complex128 and here we really want real numbers